### PR TITLE
Protect /rendered page from invalid DAG ids (#8171)

### PR DESCRIPTION
**Pull request by rcjsuen**
Redirects the user back to the main page if an invalid DAG id is provided to the `/rendered` page instead of crashing to improve the overall user experience.

This looks at one of the routes mentioned by #8171.

---

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by mik-laj** [comment]
This is what I am looking for. Can you also check other input parameters for this view i.e. ``task_id``, ``execution_date``?  If you submit incorrect values to these parameter, you should see the appropriate error messages.

**Comment by rcjsuen** [comment]
> Can you also check other input parameters for this view i.e. `task_id`, `execution_date`?

@mik-laj What kind of checks are you looking for? Just a simple `None` check or do you want exceptions caught and handled like the other code further down in the request?

https://github.com/apache/airflow/blob/45c8983306ab1c54abdacd8f870e790fad25cb37/airflow/www/views.py#L620-L628

**Comment by mik-laj** [comment]
I open following page:
```
http://localhost:28080/rendered?dag_id=example_bash_operator&task_id=runme_0&execution_date=2020-04-13T00%3A00%3A00%2B00%3A00
```
I see the correct view.
I add one letter to the task ID.
```
http://localhost:28080/rendered?dag_id=example_bash_operator&task_id=runme_0D&execution_date=2020-04-13T00%3A00%3A00%2B00%3A00
```
Screenshot 2020-04-15 at 01 26 50
I see a mushroom. This should not happen in a mature application. A clear error message should appear.

When I open following address:
```
http://localhost:28080/rendered?dag_id=example_bash_operator&task_id=runme_0&execution_date=2020-04-13T00%3A00%3A00%2B00%3A00DDDD
```
I also see mushroom.
Screenshot 2020-04-15 at 01 29 02
This is a visual problem, but it is also a security problem.  If we see similar messages, it means that we haven't verified enough input data. Data validation is the basic method of protecting against other serious attacks from the "Injection" family e.g. SQL Injection. Input validation should happen as early as possible in the data flow, preferably as soon as the data is received from the client.  However, we do not have any validation for many parameters.
image

More information:
https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html

**Comment by mik-laj** [comment]
We have not previously validated data (we have only validation provided by FAB) and we have to come up with good practises. You may have the impression that the scope of work has changed, but if we just don't start working, we don't know how to do something fully. Now I am convinced that this is an important task and we definitely should do it. Go ahead and ask if you have questions.

**Comment by rcjsuen** [comment]
Thank you for your information, @mik-laj.

1. How should I validate each of the three parameters? Is there a common utility function that can be used to check and will return a `bool`?
    - `dag_id`
    - `execution_date`
    - `task_id`
1. If the error page should **never** be displayed then it sounds to me like the entire function should be wrapped in a `try` block so that no exceptions are presented to the user by mistake. Do you agree?

**Comment by mik-laj** [comment]
1. For DAG ID and Task ID, you could use `airflow.utils.helpers.validate_key`.  For execution_date, I'm afraid there is no such function yet, but it should not be difficult to write it.
2.  We should not put the whole view in a try-catch block, but you may use the try-catch code block somewhere. We should handle all the problems we can in the code. Exceptions to the screen should only be displayed if there is no other solution. In this case, we know another solution - we can display a clear error message and redirect to the home page.

File: airflow/www/views.py [hunk 1]
```diff
@@ -603,13 +603,25 @@ def dag_details(self, session=None):
     def rendered(self):
         dag_id = request.args.get('dag_id')
         task_id = request.args.get('task_id')
+
+        dag = dagbag.get_dag(dag_id)
```
Comment (mik-laj) [inline airflow/www/views.py:607]: Can you change the order of these and next checks? We should not pass parameters to other classes if they don't have the correct format.

**Comment by mik-laj** [comment]
Have you thought to write a decorator who will perform basic check?
```
    @validate_query_params([
        'dag_id': (validate_key, 'Invalid DAG ID.')
        'task_id': (validate_key, 'Invalid DAG ID.')
    ])
    def rendered(self):
```
This will make it easier to add this in subsequent views

**Comment by mik-laj** [comment]
@rcjsuen How are you? This is a significant change because it touches on the topic of security. Let me know If there’s anything I can do to help

**Comment by rcjsuen** [comment]
> 
> 
> @rcjsuen How are you? This is a significant change because it touches on the topic of security. Let me know If there’s anything I can do to help

@mik-laj Thank you for reaching out. Apologies as I've been busy with other things. What you said made sense though.

**Comment by rcjsuen** [comment]
@mik-laj Is `airflow/utils/decorators.py` the right place to create the new decorator?

**Comment by mik-laj** [comment]
@rcjsuen Please create this decorator in airflow/www/utils.py or airflow/www/decorators.py

**Comment by rcjsuen** [comment]
Not able to invest the time this change needs and it looks like @2796gaurav may be looking at this based on https://github.com/apache/airflow/issues/8171#issuecomment-640240071.

**Pull request closed without merging**